### PR TITLE
Chore: Refactor CLIEngine tests (refs #13481)

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

Refactored a test

#### What changes did you make? (Give an overview)

I removed the in-memory file system from the `CLIEngine` tests and replaced it with an actual filesystem. This is important for working on simple config because the in-memory filesystem does all kinds of complicated things to "fake" a filesystem, and it's impossible to reuse that to test that simple config doesn't break anything.

This change keeps all of the tests working the exact same way they did previously, only with an actual filesystem. That will allow me to eventually start using exports from `@eslint/eslintrc` inside `CLIEngine` and tell if I broke anything.

I'll need to do the same for the rest of the places in the tests where in-memory filesystems are used, so this is the first of a few such refactors.

#### Is there anything you'd like reviewers to focus on?

Anything obvious look out of place?

### CODE REVIEW

File: tests/lib/cli-engine/cli-engine.js
Comment (mdjermanovic) [inline tests/lib/cli-engine/cli-engine.js:3193]: Would it be good to switch the order:

```js
cleanup = teardown.cleanup;
await teardown.prepare();
```

so the cleanup can be done even if prepare maybe partially failed, but still created some files.

File: tests/lib/cli-engine/cli-engine.js
Comment (mdjermanovic) [inline tests/lib/cli-engine/cli-engine.js:3624]: Would it be good to avoid mixing these directories with the one that gets copied in the top [before](https://github.com/eslint/eslint/blob/1df7fa69967cc514a541c363eb52d845e1b1d8f0/tests/lib/cli-engine/cli-engine.js#L83) and has its own `.eslintrc` files and `node_modules`?

File: tests/lib/cli-engine/cli-engine.js
Comment (mdjermanovic) [inline tests/lib/cli-engine/cli-engine.js:3911]: I think that `cleanup` will not delete this file.

File: tests/lib/cli-engine/cli-engine.js
Comment (mdjermanovic) [inline tests/lib/cli-engine/cli-engine.js:3930]: Same as above, this file will probably remain after the test.

**Review by mdjermanovic — COMMENTED** [review]
Looks good, I left some questions.

File: tests/lib/cli-engine/cli-engine.js
Comment (nzakas) [reply tests/lib/cli-engine/cli-engine.js:3178]: Ooh that is nicer. Thanks.

File: tests/lib/cli-engine/cli-engine.js
Comment (nzakas) [reply tests/lib/cli-engine/cli-engine.js:3193]: Done.

File: tests/lib/cli-engine/cli-engine.js
Comment (nzakas) [reply tests/lib/cli-engine/cli-engine.js:3624]: I think this is okay for now. My main goal was to do a one-to-one translation of existing tests. I'd rather not get into changing how they work at this point.

File: tests/lib/cli-engine/cli-engine.js
Comment (nzakas) [reply tests/lib/cli-engine/cli-engine.js:3911]: Yeah, nice catch. Let me look at that.

Comment (nzakas) [reply tests/lib/cli-engine/cli-engine.js:3911]: I verified that this file is still removed. The library does a `path.resolve()` for each file and uses the absolute path for operations, so it doesn't restrict to the root.

File: tests/lib/cli-engine/cli-engine.js
Comment (nzakas) [reply tests/lib/cli-engine/cli-engine.js:3930]: Same, this file is safely removed.

**Comment by nzakas** [comment]
Feedback addressed.

File: tests/lib/cli-engine/cli-engine.js
Comment (mdjermanovic) [reply tests/lib/cli-engine/cli-engine.js:3932]: Hm, I tried with commenting out this: 

https://github.com/eslint/eslint/blob/1df7fa69967cc514a541c363eb52d845e1b1d8f0/tests/lib/cli-engine/cli-engine.js#L92-L94

and the file was still there, so it looks like it remains undeleted between tests.

I think the lib removes only the innermost directory that was initally passed as `baseDir`:

https://github.com/kettanaito/fs-teardown/blob/b801c07f764563cd413685c523dd9319247c720e/src/createTeardown.ts#L80-L82

Comment (mdjermanovic) [reply tests/lib/cli-engine/cli-engine.js:3932]: Another way to check, try to completely remove this test and then run all tests:

https://github.com/eslint/eslint/blob/db6787f7478772d1c26d93fbba041dcde4a23585/tests/lib/cli-engine/cli-engine.js#L3928-L3945

If we just remove this test (or change the order of tests), the invalid config file `"../.eslintrc.json": "BROKEN FILE"` from the previous test will not be overwritten with a valid one from this test, so all of the remaining tests will fail because they'll have a broken config file in an ancestor directory which wasn't cleaned up.

File: tests/lib/cli-engine/cli-engine.js
Comment (btmills) [inline tests/lib/cli-engine/cli-engine.js:3673]: This is one of a few other places that could be updated to the [refactor @mdjermanovic suggested](https://github.com/eslint/eslint/pull/13709#discussion_r495894365). Did you want to do that? Looks like a search for `afterEach(async` in the file should find the rest if you decide to change them.

File: tests/lib/cli-engine/cli-engine.js
Comment (btmills) [inline tests/lib/cli-engine/cli-engine.js:3180]: In other places you use the simpler `afterEach(cleanup);` form, but this one was left as an explicit `async` after you removed the `cleanup = () => { };` line from its body in db6787f7478772d1c26d93fbba041dcde4a23585.

```suggestion
            afterEach(cleanup);
```

File: tests/lib/cli-engine/cli-engine.js
Comment (btmills) [inline tests/lib/cli-engine/cli-engine.js:3928]:
```suggestion
            it.skip("'executeOnFiles(\".\")' should not ignore '.' even if 'ignorePatterns' contains it.", async () => {
```

I'm also seeing that the broken `../.eslintrc.json` file isn't cleaned up at the end of the previous test as @mdjermanovic found. It's only cleaned up when the whole temp fixture directory is deleted at the very end. If you add a `.skip` to this test, all the remaining tests that don't specify `root: true` will fail.

One solution could be to add another level of directory nesting to `root` for the two `../.eslintrc.json` tests so that the outermost `.eslintrc.json` is still inside the teardown directory.

**Review by btmills — CHANGES_REQUESTED** [review]
I saw two inconsistencies in your `beforeEach`/`afterEach` patterns. Not sure if that falls under the category of looking out of place, so I'll leave it to you to decide how much you care that they all follow the same pattern.

I left a code suggestion for a `.skip` that will demonstrate the broken `../.eslintrc.json` not being cleaned up after its test. Hopefully that's enough to help you diagnose. Once that's resolved, LGTM.

File: tests/lib/cli-engine/cli-engine.js
Comment (nzakas) [reply tests/lib/cli-engine/cli-engine.js:3929]: I really don't want to change how the current tests operate, as they are already complicated enough. I think the easiest solution is just to remove the extra file as part of the cleanup.

**Comment by nzakas** [comment]
Updated to:

1. Explicitly delete the `../.eslintrc.json` file in the one block where that's used.
2. Normalize the `beforeEach`/`afterEach` patterns. Note that in cases where `cleanup` is being overwritten inside a test case, I needed to use `afterEach(() => cleanup())` instead of `afterEach(cleanup)` because the latter evaluates `cleanup` immediately (resulting in `undefined`) and the former evaluates `cleanup` at the time it executes.

**Review by btmills — APPROVED** [review]
Works for me! LGTM

File: tests/lib/cli-engine/cli-engine.js
Comment (mdjermanovic) [inline tests/lib/cli-engine/cli-engine.js:3193]: Should we switch the order like this (to set `cleanup` before doing `prepare()`) in other tests as well, for consistency (there are 19 tests with this pattern, can be found by searching for `await teardown.prepare()`)

File: tests/lib/cli-engine/cli-engine.js
Comment (mdjermanovic) [inline tests/lib/cli-engine/cli-engine.js:3907]: The last test doesn't create `"../.eslintrc.json"`, so `shell.rm` will output something like this when we run `npm test`:

```
rm: no such file or directory: /tmp/eslint/fixtures/cli-engine/.eslintrc.json
```

(observable on CI, too)

**Review by mdjermanovic — COMMENTED** [review]
Just two small notes. LGTM if you think there's no need to fix those at the moment.

File: tests/lib/cli-engine/cli-engine.js
Comment (nzakas) [reply tests/lib/cli-engine/cli-engine.js:3907]: Okay that's weird, I wasn't seeing this locally. I'll fix it.

File: tests/lib/cli-engine/cli-engine.js
Comment (nzakas) [reply tests/lib/cli-engine/cli-engine.js:3193]: I don't think this is important enough to spend time fixing. Let's focus on higher level issues and keep in mind this file will be going away once we remove `CLIEngine`.

**Review by btmills — APPROVED** [review]

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
